File: algorithms/03_cpp/strings/aho_corasick.cpp
// Aho-Corasick multi-pattern matching.
#include <algorithm>
#include <array>
#include <cassert>
#include <iostream>
#include <queue>
#include <string>
#include <vector>

struct Node {
    std::array<int, 26> next;
    int fail = 0;
    std::vector<std::string> out;
    Node() { next.fill(-1); }
};

class AhoCorasick {
  public:
    AhoCorasick() { nodes_.emplace_back(); }

    void add(const std::string &pattern) {
        int node = 0;
        for (char ch : pattern) {
            int c = ch - 'a';
            if (nodes_[node].next[c] == -1) {
                nodes_[node].next[c] = static_cast<int>(nodes_.size());
                nodes_.emplace_back();
            }
            node = nodes_[node].next[c];
        }
        nodes_[node].out.push_back(pattern);
    }

    void build() {
        std::queue<int> queue;
        for (int c = 0; c < 26; ++c)
            if (nodes_[0].next[c] != -1) { nodes_[nodes_[0].next[c]].fail = 0; queue.push(nodes_[0].next[c]); }
        while (!queue.empty()) {
            int node = queue.front();
            queue.pop();
            for (int c = 0; c < 26; ++c) {
                int child = nodes_[node].next[c];
                if (child != -1) {
                    int fallback = nodes_[node].fail;
                    while (fallback && nodes_[fallback].next[c] == -1) fallback = nodes_[fallback].fail;
                    if (nodes_[fallback].next[c] != -1) fallback = nodes_[fallback].next[c];
                    nodes_[child].fail = fallback;
                    nodes_[child].out.insert(nodes_[child].out.end(),
                                             nodes_[fallback].out.begin(), nodes_[fallback].out.end());
                    queue.push(child);
                }
            }
        }
    }

    std::vector<std::string> search(const std::string &text) {
        std::vector<std::string> found;
        int node = 0;
        for (char ch : text) {
            int c = ch - 'a';
            while (node && nodes_[node].next[c] == -1) node = nodes_[node].fail;
            node = nodes_[node].next[c] == -1 ? 0 : nodes_[node].next[c];
            found.insert(found.end(), nodes_[node].out.begin(), nodes_[node].out.end());
        }
        return found;
    }

  private:
    std::vector<Node> nodes_;
};

int main() {
    AhoCorasick machine;
    for (const std::string &pattern : {"he", "she", "his", "hers"}) machine.add(pattern);
    machine.build();
    auto found = machine.search("ushers");
    assert(std::find(found.begin(), found.end(), "he") != found.end());
    assert(std::find(found.begin(), found.end(), "she") != found.end());
    std::cout << "aho-corasick ok\n";
    return 0;
}

File: algorithms/07_typescript/strings/aho_corasick.ts
/**
 * Aho-Corasick Automaton
 * Multi-pattern dictionary-matching algorithm with linear construction O(sum(m)) and search O(n + z).
 */

class ACNode {
  children = new Map<string, ACNode>();
  fail: ACNode | null = null;
  output: string[] = [];
}

export class AhoCorasick {
  root = new ACNode();

  constructor(patterns: string[]) {
    this.buildTrie(patterns);
    this.buildFailureLinks();
  }

  private buildTrie(patterns: string[]): void {
    for (const pat of patterns) {
      let curr = this.root;
      for (const ch of pat) {
        if (!curr.children.has(ch)) {
          curr.children.set(ch, new ACNode());
        }
        curr = curr.children.get(ch)!;
      }
      curr.output.push(pat);
    }
  }

  private buildFailureLinks(): void {
    const queue: ACNode[] = [];
    for (const child of this.root.children.values()) {
      child.fail = this.root;
      queue.push(child);
    }

    while (queue.length > 0) {
      const curr = queue.shift()!;
      for (const [ch, child] of curr.children.entries()) {
        let f = curr.fail;
        while (f && !f.children.has(ch)) f = f.fail;
        child.fail = f ? f.children.get(ch)! : this.root;
        child.output.push(...child.fail.output);
        queue.push(child);
      }
    }
  }

  search(text: string): { pattern: string; index: number }[] {
    const matches: { pattern: string; index: number }[] = [];
    let curr: ACNode | null = this.root;

    for (let i = 0; i < text.length; i++) {
      const ch = text[i];
      while (curr && !curr.children.has(ch)) curr = curr.fail;
      curr = curr ? curr.children.get(ch)! : this.root;

      for (const pat of curr.output) {
        matches.push({ pattern: pat, index: i - pat.length + 1 });
      }
    }
    return matches;
  }
}

const ac = new AhoCorasick(["he", "she", "his", "hers"]);
const acMatches = ac.search("ahishers");
if (acMatches.length !== 3) throw new Error("Aho-Corasick failed");
console.log("Aho-Corasick Automaton verified successfully.");
